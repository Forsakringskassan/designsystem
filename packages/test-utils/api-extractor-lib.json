{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/lib.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/lib.d.ts"
    }
}
