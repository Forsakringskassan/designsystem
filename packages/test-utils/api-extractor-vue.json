{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/vue/index.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/vue.d.ts"
    }
}
