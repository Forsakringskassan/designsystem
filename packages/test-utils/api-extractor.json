{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "tsdocMetadata": {
        "tsdocMetadataFilePath": "<projectFolder>/dist/types/tsdoc-metadata.json"
    },
    "dtsRollup": {
        "enabled": false
    }
}
