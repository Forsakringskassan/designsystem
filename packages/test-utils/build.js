const fs = require("node:fs/promises");
const path = require("node:path");
const esbuild = require("esbuild");

const type = {
    cjs: "commonjs",
    esm: "module",
};

async function build() {
    for (const format of ["cjs", "esm"]) {
        const outdir = `dist/${format}`;
        const common = {
            logLevel: "info",
            sourcemap: true,
            external: ["vue", "@fkui/*"],
            bundle: true,
            platform: "node",
            format,
            target: "node20",
            outdir,
        };
        await esbuild.build({
            ...common,
            entryPoints: ["src/lib.ts", "src/jest.ts"],
        });
        await esbuild.build({
            ...common,
            entryPoints: [{ in: "src/vue/index.ts", out: "vue" }],
            define: {
                "process.env.NODE_ENV": JSON.stringify("development"),
            },
        });
        const pkg = JSON.stringify({ type: type[format] }, null, 2);
        await fs.writeFile(path.join(outdir, "package.json"), pkg, "utf-8");
    }
}

build().catch((err) => {
    console.error(err);
});
