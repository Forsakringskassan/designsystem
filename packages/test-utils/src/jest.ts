/* imported for side-effects only */
import "./matchers";

export * from "./lib";
