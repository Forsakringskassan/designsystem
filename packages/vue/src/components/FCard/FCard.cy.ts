import { type DefineComponent, defineComponent } from "vue";
import {
    DensityWrapper,
    densityWrapperHeight,
    densityWrapperWidth,
} from "@fkui/test-utils/vue";
import { FTextField } from "../FTextField";
import FCard from "./FCard.vue";
import FCardValidationExample from "./examples/FCardValidationExample.vue";

function createComponent(): DefineComponent {
    return defineComponent({
        template: /* HTML */ `
            <f-card>
                <template #header="{ headingSlotClass }">
                    <h3 :class="headingSlotClass">Rubrik</h3>
                </template>
                <template #default>
                    <f-text-field
                        v-validation.maxLength="{ maxLength: { length: 100 } }"
                    >
                        Förnamn
                    </f-text-field>
                    <f-text-field
                        v-validation.maxLength="{ maxLength: { length: 100 } }"
                    >
                        Efternamn
                    </f-text-field>
                </template>
                <template #footer>
                    <div class="button-group">
                        <button
                            class="button button-group__item button--discrete"
                            type="button"
                        >
                            <span> Ta bort </span>
                        </button>

                        <button
                            class="button button-group__item button--discrete"
                            type="button"
                        >
                            <span> Ändra </span>
                        </button>
                    </div>
                </template>
            </f-card>
        `,
        components: {
            FCard,
            FTextField,
        },
    });
}

describe("validation", () => {
    it("should show error when invalid", () => {
        cy.mount(FCardValidationExample);
        cy.get(`[data-test="set-invalid"]`).click();
        cy.get(".card").should("contain.text", "Uppgifter saknas");
    });

    it("should set focus at submit", () => {
        cy.mount(FCardValidationExample);
        cy.get(`[data-test="set-invalid"]`).click();
        cy.get(`[type="submit"]`).click();
        cy.get(`#focus`).should("have.focus");
    });

    it("should not throw error when child component is validated", () => {
        cy.mount(
            defineComponent({
                template: /* HTML */ `
                    <f-card>
                        <template #default>
                            <f-text-field
                                v-validation.maxLength="{ maxLength: { length: 100 } }"
                            >
                                Etikett
                            </f-text-field>
                        </template>
                    </f-card>
                `,
                components: {
                    FCard,
                    FTextField,
                },
            }),
        );
        cy.get("input").focus();
        cy.get("input").blur();

        // eslint-disable-next-line cypress/no-unnecessary-waiting -- required for potential error to be thrown and fail test
        cy.wait(1);
    });
});

/* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
describe.skip("Screenshot", () => {
    beforeEach(() => {
        cy.viewport(1024, 600);
    });
    it("should match screenshot", () => {
        cy.mount(createComponent());
        cy.get(".card").toMatchScreenshot();
    });
});

/* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
describe.skip("density", () => {
    const DensityComponent = defineComponent({
        template: /* HTML */ `
            <density-wrapper>
                <f-card>
                    <template #header="{ headingSlotClass }">
                        <h3 :class="headingSlotClass">Rubrik</h3>
                    </template>
                    <template #default> Innehåll </template>
                    <template #footer>
                        <div class="button-group">
                            <button
                                class="button button-group__item button--tertiary button--medium button--align-text"
                                type="button"
                            >
                                <f-icon name="pen"></f-icon>
                                <span> Ändra </span>
                            </button>
                            <button
                                class="button button-group__item button--tertiary button--medium button--align-text"
                                type="button"
                            >
                                <f-icon name="trashcan"></f-icon>
                                <span> Ta bort </span>
                            </button>
                        </div>
                    </template>
                </f-card>
            </density-wrapper>
        `,
        components: {
            DensityWrapper,
            FCard,
        },
    });

    it(`should be densified`, () => {
        cy.viewport(densityWrapperWidth, densityWrapperHeight);
        cy.mount(DensityComponent);
        cy.toMatchScreenshot();
    });
});
