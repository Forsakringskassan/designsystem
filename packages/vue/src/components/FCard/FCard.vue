<template>
    <div class="card card--default">
        <div v-if="$slots.header" class="card__header">
            <!--@slot Slot for the title. -->
            <slot name="header" v-bind="{ headingSlotClass: headingClass }"></slot>
        </div>
        <div class="card__content">
            <!--@slot Slot for the main content, e.g. paragraphs, input fields, etc. -->
            <slot name="default"></slot>
        </div>
        <div v-if="$slots.footer" class="card__footer">
            <!--@slot Slot the footer content, i.e. buttons. -->
            <slot name="footer"></slot>
        </div>
    </div>
</template>

<script setup lang="ts">
import { ref } from "vue";

const headingClass = ref(["card__header-label"]);
</script>
