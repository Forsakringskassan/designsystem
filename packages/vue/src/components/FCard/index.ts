export { default as FCard } from "./FCard.vue";
