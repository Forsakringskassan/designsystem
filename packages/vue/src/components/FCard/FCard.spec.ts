import { VueWrapper, mount } from "@vue/test-utils";
import {
    cjsResolver,
    FileSystemConfigLoader,
    HtmlValidate,
} from "html-validate/node";
import FCard from "./FCard.vue";
import "html-validate/jest";

const loader = new FileSystemConfigLoader([cjsResolver()], {
    extends: [
        "html-validate:recommended",
        "html-validate-vue:recommended",
        "@fkui/vue:recommended",
    ],
    plugins: ["<rootDir>/htmlvalidate", "html-validate-vue"],
});
const htmlvalidate = new HtmlValidate(loader);

function createWrapper({
    props = {},
    listeners = {},
    attrs = {},
} = {}): VueWrapper {
    return mount(FCard, {
        attrs: { ...attrs },
        props: { ...props },
        slots: {
            header: "Header slot",
            default: "Content slot",
            footer: "Footer slot",
        },
        listeners: { ...listeners },
    });
}

describe("snapshots", () => {
    it("should match snapshot", () => {
        const wrapper = createWrapper();
        expect(wrapper.element).toMatchSnapshot();

        expect(wrapper.get(".card__header")).toBeTruthy();
        expect(wrapper.get(".card__footer")).toBeTruthy();
    });

    it("should not render header class when heading slot is omitted", () => {
        const wrapper = mount(FCard);
        expect(wrapper.find(".card__header").exists()).toBeFalsy();
    });

    it("should not render footer class when footer slot is omitted", () => {
        const wrapper = mount(FCard);
        expect(wrapper.find(".card__footer").exists()).toBeFalsy();
    });
});

describe("html-validate", () => {
    it("should allow valid values", async () => {
        expect.assertions(1);
        const markup = /* HTML */ `
            <f-card>
                <template #default> Content </template>
            </f-card>
            <f-card>
                <template #header></template>
                <template #default> Content </template>
                <template #footer></template>
            </f-card>
        `;
        const report = await htmlvalidate.validateString(markup);
        expect(report).toBeValid();
    });

    it("should not allow invalid values", async () => {
        expect.assertions(1);
        const markup = /* HTML */ ` <f-card></f-card> `;
        const report = await htmlvalidate.validateString(markup);
        expect(report).toMatchInlineCodeframe(`
            "error: <f-card> component requires slot "default" to be implemented (vue/required-slots) at inline:1:3:
            > 1 |  <f-card></f-card>
                |   ^^^^^^
            Selector: f-card"
        `);
    });
});
