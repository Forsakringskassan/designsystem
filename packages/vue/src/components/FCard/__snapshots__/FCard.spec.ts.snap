// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot 1`] = `
<div
  class="card card--default"
  id="fkui-vue-element-0001"
>
  <div
    class="card__header"
  >
    
    Header slot
    
  </div>
  
  <!--v-if-->
  
  <div
    class="card__content"
  >
    
    Content slot
    
  </div>
  <div
    class="card__footer"
  >
    
    Footer slot
    
  </div>
</div>
`;
